gridappsd/simulation.py

```diff
@@ -1,63 +1,140 @@
+
+from copy import deepcopy
 import json
-import sys
-import time
+import logging
+
+from . import topics as t
+from .topics import simulation_input_topic
 
-from . topics import simulation_input_topic, simulation_output_topic, simulation_log_topic
+_log = logging.getLogger(__name__)
 
 
 class Simulation(object):
-    def __init__(self, gapps, simulation_id: str, duration: int, timestep_finished=None):
-        self.gappsd = gapps
-        self.simulation_id = simulation_id
-        self._show_timesteps = True
-        self._duration = int(duration)
-        self._timestep_requested = 0
-        self._timestep_finished_callback = timestep_finished
-        self._is_running = True
-        self.gappsd.subscribe(simulation_log_topic(simulation_id), self._on_message)
-
-    def _on_message(self, headers, message):
-        if 'logMessage' in message:
-            log_message = message['logMessage']
-            if log_message.startswith("done with timestep"):
-                timestep = int(log_message[len("done with timestep "):])
-                if self._timestep_finished_callback is not None:
-                    self._timestep_finished_callback(self, timestep)
-                # We use -1 here because the last timestep from fncs-goss-bridge.py never actually
-                # comes through at least as of 2019-6-7
-                if timestep >= self._duration-1:
-                    self._is_running = False
-
-        if self._show_timesteps:
-            # print(headers, message)
-            # print(headers)
-            sys.stdout.write(f"{message['logMessage']}\n")
-            # print(message["logMessage"])
-
-    @property
-    def running(self):
-        return self.is_running()
-
-    @property
-    def showing_status(self):
-        return self._show_timesteps
+    """ Simulation object allows controlling simulations through a python API.
+
+    The simulation object allows controlling and monitoring of simulations through
+    a python API.  It is capable of starting, stopping, pausing and restarting simulations
+    that are run on GridAPPSD.
+
+    There are three events that can be monitored by multiple callbacks:
+
+        - timestep:
+
+            This event is triggered each time a timestep has been triggered
+
+        - on_timestep_begin:
+
+            This event is triggered when a new tiemtep has been started.
+
+        - on_finished:
+
+            This event is triggered when the simulation has completed.
+
+    """
+    def __init__(self, gapps, run_config):
+        # Protect from circular import.
+        from . import GridAPPSD
+        assert isinstance(gapps, GridAPPSD), "Must be an instance of GridAPPSD"
+        assert isinstance(run_config, dict)
+        self._gapps = gapps
+        self._run_config = deepcopy(run_config)
+
+        # Will be populated when the simulation is first started.
+        self.simulation_id = None
+
+        self.__on_start = set()
```

Comment:
@craig8 - any particular reason for using sets instead of lists? Why can't I call a function twice? I don't immediately see a use case for calling a function multiple times, but I don't see why we should disallow it.